pkg/controller/workload/jobframework/helpers.go

```diff
@@ -28,14 +29,180 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/constants"
 	utilpriority "sigs.k8s.io/kueue/pkg/util/priority"
 	"sigs.k8s.io/kueue/pkg/workload"
 )
 
-// EnsureOneWorkload will query for the single matched workload corresponding to job and return it.
+// GenericJobReconciler reconciles a GenericJob object
+type GenericJobReconciler struct {
+	client                     client.Client
+	scheme                     *runtime.Scheme
+	record                     record.EventRecorder
+	manageJobsWithoutQueueName bool
+	waitForPodsReady           bool
+}
+
+type options struct {
+	manageJobsWithoutQueueName bool
+	waitForPodsReady           bool
+}
+
+// Option configures the reconciler.
+type Option func(*options)
+
+// WithManageJobsWithoutQueueName indicates if the controller should reconcile
+// jobs that don't set the queue name annotation.
+func WithManageJobsWithoutQueueName(f bool) Option {
+	return func(o *options) {
+		o.manageJobsWithoutQueueName = f
+	}
+}
+
+// WithWaitForPodsReady indicates if the controller should add the PodsReady
+// condition to the workload when the corresponding job has all pods ready
+// or succeeded.
+func WithWaitForPodsReady(f bool) Option {
+	return func(o *options) {
+		o.waitForPodsReady = f
+	}
+}
+
+var defaultOptions = options{}
+
+func NewReconciler(
+	scheme *runtime.Scheme,
+	client client.Client,
+	record record.EventRecorder,
+	opts ...Option) *GenericJobReconciler {
+	options := defaultOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	return &GenericJobReconciler{
+		scheme:                     scheme,
+		client:                     client,
+		record:                     record,
+		manageJobsWithoutQueueName: options.manageJobsWithoutQueueName,
+		waitForPodsReady:           options.waitForPodsReady,
+	}
+}
+
+func (r *GenericJobReconciler) Reconcile(ctx context.Context, req ctrl.Request, job GenericJob) (ctrl.Result, error) {
+	namespacedName := types.NamespacedName{Name: job.Object().GetName(), Namespace: job.Object().GetNamespace()}
+
+	log := ctrl.LoggerFrom(ctx).WithValues("job", namespacedName.String())
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	isStandaloneJob := job.ParentWorkloadName() == ""
+
+	// when manageJobsWithoutQueueName is disabled we only reconcile jobs that have either
+	// queue-name or the parent-workload annotation set.
+	if !r.manageJobsWithoutQueueName && job.QueueName() == "" && isStandaloneJob {
+		log.V(3).Info(fmt.Sprintf("Neither %s, nor %s annotation is set, ignoring the job", constants.QueueAnnotation, constants.ParentWorkloadAnnotation))
+		return ctrl.Result{}, nil
+	}
+
+	log.V(2).Info("Reconciling Job")
```

Comment:
Please share how these log lines look. In particular, I would like to know if it's clear when the log refers to a job or an mpijob. I think the log name should cover it.

Reply:
This is an example log line for MPI Job: 
```
2023-03-14T10:03:29.470733346+01:00   LEVEL(-2)       jobframework/helpers.go:112     Reconciling Job {"controller": "mpijob", "controllerGroup": "kubeflow.org", "controllerKind": "MPIJob", "MPIJob": {"name":"test-job","namespace":"default"}, "namespace": "default", "name": "test-job", "reconcileID": "8dbde845-2ccf-4341-8fd7-749225243656", "job": "default/test-job"}
```
and for Batch Job:
```
  2023-03-14T09:43:40.691102798+01:00   LEVEL(-2)       jobframework/helpers.go:112     Reconciling Job {"controller": "job", "controllerGroup": "batch", "controllerKind": "Job", "Job": {"name":"test-job","namespace":"default"}, "namespace": "default", "name": "test-job", "reconcileID": "414137e7-60b0-4edc-9ff1-753b81ab2186", "job": "default/test-job"}
  ```